arch/armv7m/jmp.S

```diff
@@ -7,91 +7,116 @@
  * _setjmp, _longjmp, setjmp, sigsetjmp
  *
  * Copyright 2018, 2019 Phoenix Systems
- * Author: Kamil Amanowicz, Aleksander Kaminski
+ * Author: Kamil Amanowicz, Aleksander Kaminski, Jakub Sarzyński
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
+ /* setjmp env buffer description
+  * Offset	Description
+  * 0		Signal mask flag - if != 0 then next value stores a signal mask
+  * 4		Signal mask
+  * 8		SP register
+  * 12-44	r4-r11 registers
+  * 44		LR register
+  * 48-80	d8-d15 registers
+  * 80		fpscr register
+  */
+
 #define __ASSEMBLY__
 
 .thumb
 .syntax unified
 
-.globl _setjmp
-.type _setjmp, %function
-_setjmp:
-	/* set jump address and stack pointer */
-	ldr r1, =(1f + 1)
-	mov r2, sp
-	/* store everythig in jmpbuf */
-	stm r0, {r1-r2, r4-r11, lr}
+
+.globl sigsetjmp
+.type sigsetjmp, %function
+sigsetjmp:
+	/* Store signal mask if savesigs != 0 */
+	mov r2, #0
+	cmp r1, #0
+	beq 1f
+
+	/* Get signal mask */
+	sub sp, #4
+	push {r0, r1, lr}
 	mov r0, #0
-	bx lr
+	mov r1, #0
+	bl signalMask
+	mov r2, r0
+	pop {r0, r1, lr}
+	add sp, #4
+
 1:
-	/* this is where we land after the jump
-	 * just pass the return value which is in r3 */
-	mov r0, r3
+	/* Store mask (fake or real) */
+	stmia r0!, {r1, r2}
+
+	/* Store general registers */
+	mov r2, sp
+	stmia r0!, {r2, r4-r11, lr}
+
+#ifndef __SOFTFP__
+	/* Store fpu registers */
+	vstm r0!, {d8-d15}
+	vmrs r2, fpscr
+	str r2, [r0], #4
+#endif
+
+	mov r0, #0
 	bx lr
+.size sigsetjmp, .-sigsetjmp
+
+.globl _setjmp
+.type _setjmp, %function
+_setjmp:
+	mov r1, #0
+	b sigsetjmp
 .size _setjmp, .-_setjmp
 
+.globl setjmp
+.type setjmp, %function
+setjmp:
+	mov r1, #1
+	b sigsetjmp
+.size setjmp, .-setjmp
+
+
 .globl _longjmp
 .type _longjmp, %function
 _longjmp:
-	/* move return value to r3.
-	 * longjmp should not make setjmp return 0 though */
-	cmp r1, #0
-	ite eq
-	moveq r3, #1
-	movne r3, r1
-	/* restore registers from jmpbuf */
-	ldmia r0!, {r1-r2, r4-r11, lr}
-	mov sp, r2
-	bx r1
-.size _longjmp, .-_longjmp
+	/* Check for signal mask in env buffer, if it's saved restore it */
+	ldmia r0!, {r2-r3}
+	cmp r2, #0
+	beq 1f
 
-.globl setjmp
-.type setjmp, %function
-setjmp:
-	push {r0, lr}
-	/* get signal mask */
-	mov r0, #0
-	mov r1, #0
+	/* Restore mask */
+	sub sp, #4
+	push {r0, r1, lr}
```

Comment:
Is it necessary to preserve the `lr` register? In line 105 the current value is overwriten.